File: MessageBuffer.js
class MessageBuffer {
    constructor() {
        this.buffer = Buffer.alloc(0);
        this.onDataCallback = null;
    }

    setOnDataCallback(callback) {
        this.onDataCallback = callback;
    }

    consume(data) {
        // Append new data to existing buffer
        this.buffer = Buffer.concat([this.buffer, data]);

        while (this.buffer.length >= 4) {
            // Read the size prefix (first 4 bytes)
            const size = this.buffer.readUInt32BE(0);

            // Check if we have received the complete frame
            if (this.buffer.length >= size + 4) {
                // Extract the complete frame (excluding size prefix)
                const frame = this.buffer.slice(4, size + 4);
                
                // Remove the processed frame from buffer
                this.buffer = this.buffer.slice(size + 4);

                // Call the callback with the complete frame
                if (this.onDataCallback) {
                    this.onDataCallback(frame);
                }
            } else {
                // Not enough data for a complete frame, wait for more
                break;
            }
        }
    }

    produce(data) {
        // Ensure data is a Buffer
        const dataBuffer = Buffer.isBuffer(data) ? data : Buffer.from(data);
        
        // Create a buffer for the size prefix (4 bytes)
        const sizeBuffer = Buffer.alloc(4);
        // Write the size of the data as a 32-bit big-endian integer
        sizeBuffer.writeUInt32BE(dataBuffer.length, 0);
        
        // Combine size prefix and data
        return Buffer.concat([sizeBuffer, dataBuffer]);
    }
}

module.exports = MessageBuffer; 

File: go-reverse-proxy/message_buffer.go
package main

import (
	"bytes"
	"encoding/binary"
)

// MessageBuffer handles message framing and buffering
type MessageBuffer struct {
	buffer bytes.Buffer
	onData func([]byte)
}

// NewMessageBuffer creates a new MessageBuffer instance
func NewMessageBuffer() *MessageBuffer {
	return &MessageBuffer{
		buffer: bytes.Buffer{},
	}
}

// SetOnDataCallback sets the callback function for when a complete message is received
func (mb *MessageBuffer) SetOnDataCallback(callback func([]byte)) {
	mb.onData = callback
}

// Consume processes incoming data and extracts complete messages
func (mb *MessageBuffer) Consume(data []byte) {
	mb.buffer.Write(data)

	for {
		// Check if we have enough data for the length prefix
		if mb.buffer.Len() < 4 {
			return
		}

		// Read the length prefix
		lengthBytes := mb.buffer.Bytes()[:4]
		length := binary.BigEndian.Uint32(lengthBytes)

		// Check if we have the complete message
		if mb.buffer.Len() < int(length)+4 {
			return
		}

		// Extract the message
		message := make([]byte, length)
		mb.buffer.Read(lengthBytes) // Skip the length prefix
		mb.buffer.Read(message)

		// Process the message
		if mb.onData != nil {
			mb.onData(message)
		}
	}
}

// Produce creates a framed message with length prefix
func (mb *MessageBuffer) Produce(data []byte) []byte {
	length := uint32(len(data))
	lengthBytes := make([]byte, 4)
	binary.BigEndian.PutUint32(lengthBytes, length)

	// Combine length prefix and message
	result := make([]byte, 0, len(lengthBytes)+len(data))
	result = append(result, lengthBytes...)
	result = append(result, data...)

	return result
}
